tools/cell_census_builder/validate.py

```diff
@@ -255,56 +256,80 @@ def validate_axis_dataframes(
+
+
 def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[ExperimentSpecification]]) -> bool:
     """
     Validate that a single dataset is correctly represented in the census.
     Intended to be dispatched from validate_X_layers.
 
     Currently, implements weak tests:
+    * the sum is correct
     * the nnz is correct
     * there are no zeros explicitly saved (this is mandated by cell census schema)
+    * the number of present genes in each h5ad dataset == the row sum of the corresponding datasets in the presence
+      matrix.
     """
     assets_path, soma_path, dataset, experiment_specifications = args
     _, unfiltered_ad = next(open_anndata(assets_path, [dataset]))
 
     for eb in experiment_specifications:
+        error = f"{eb.name}:{dataset.dataset_id}" + " 'raw' {} mismatch {} vs {}"
         with open_experiment(soma_path, eb) as exp:
             anndata_cell_filter = make_anndata_cell_filter(eb.anndata_cell_filter_spec)
             ad = anndata_cell_filter(unfiltered_ad, retain_X=True)
-
-            soma_joinids: npt.NDArray[np.int64] = (
-                exp.obs.read(
-                    column_names=["soma_joinid", "dataset_id"], value_filter=f"dataset_id == '{dataset.dataset_id}'"
-                )
-                .concat()
-                .to_pandas()
-                .soma_joinid.to_numpy()
-            )
-
-            raw_nnz = 0
-            if len(soma_joinids) > 0:
-                assert soma.SparseNDArray.exists(exp.ms["RNA"].X["raw"].uri)
-
-                def count_elements(arr: soma.SparseNDArray, join_ids: npt.NDArray[np.int64]) -> int:
-                    return sum(t.non_zero_length for t in arr.read((join_ids, slice(None))).coos())
-
-                raw_nnz = count_elements(exp.ms["RNA"].X["raw"], soma_joinids)
-
-            def count_nonzero(arr: Union[sparse.spmatrix, npt.NDArray[Any]]) -> int:
-                """Return _actual_ non-zero count, NOT the stored value count."""
-                if isinstance(arr, (sparse.spmatrix, sparse.coo_array, sparse.csr_array, sparse.csc_array)):
-                    return np.count_nonzero(arr.data)
-                return np.count_nonzero(arr)
-
             if ad.raw is None:
-                assert raw_nnz == count_nonzero(
-                    ad.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.X)}"
+                expected_ad_x = ad.X
             else:
-                assert raw_nnz == count_nonzero(
-                    ad.raw.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.raw.X)}"
-
+                expected_ad_x = ad.raw.X
+
+            with exp.axis_query(
+                measurement_name="RNA",
+                obs_query=soma.AxisQuery(value_filter=f"dataset_id == '{dataset.dataset_id}'"),
+                var_query=soma.AxisQuery(value_filter=f"feature_id in {[*ad.var.index.array]}"),
```

Comment:
> It will also be faster to find the dataset's soma_joinid from census['census_summary']['census_datasets'] and pass that in for obs_query= via coords and not value_filter

This will the accuracy of this check to the accuracy of the  `census['census_summary']`. The `value_filter` while slower is less dependent on previous assumptions.

Reply:
Fair concern. Though we could validate separately that a dataset soma_joinid return all and only rows for the given datasets (maybe we already are somewhere?).

Reply:
Removing the var value_filter, speed up the tests significantly